src/gui/activityjob.cpp

```diff
@@ -0,0 +1,147 @@
+    activity._objectType = activityJson.value("object_type").toString();
+    activity._accName = _account->displayName();
+    activity._id = activityJson.value("activity_id").toInt();
+    activity._fileAction = activityJson.value("type").toString();
+    activity._subject = activityJson.value("subject").toString();
+    activity._message = activityJson.value("message").toString();
+    activity._file = activityJson.value("object_name").toString();
+    activity._link = QUrl(activityJson.value("link").toString());
+    activity._dateTime = QDateTime::fromString(activityJson.value("datetime").toString(), Qt::ISODate);
+    activity._icon = activityJson.value("icon").toString();
+
+    return activity;
+}
+
+std::vector<Activity> OcsActivityJob::jsonArrayToActivities(const QJsonArray &activitiesJson)
+{
+    std::vector<Activity> activities;
+    for (const auto activ : activitiesJson) {
+        activities.push_back(jsonObjectToActivity(activ.toObject()));
+    }
+    return activities;
+}
+
+void OcsActivityJob::processActivities(const QJsonDocument &json)
+{
+    const auto activitiesJson = json.object().value(QStringLiteral("ocs")).toObject().value(QStringLiteral("data")).toArray();
+    emit finished(jsonArrayToActivities(activitiesJson));
+}
+
+static QUrl parseLinkFromHeader(const QByteArray &linkRawHeader)
+{
+    // The link is ecapsulated in a tag. We need to parse this tag.
+
+    constexpr auto linkHeaderTagMinimumExpectedLength = 14;
+    if (linkRawHeader.size() <= linkHeaderTagMinimumExpectedLength) {
+        qCWarning(lcActivityJob) << "Link to next page submitted, but could not be parsed";
+        return {};
+    }
+
+    constexpr auto linkHeaderTagUnnecessaryExcessPrefixLength = 1;
+    constexpr auto linkHeaderTagUnnecessaryExcessSuffixLength = 14;
+    return QString(linkRawHeader.chopped(linkHeaderTagUnnecessaryExcessSuffixLength)
+                       .mid(linkHeaderTagUnnecessaryExcessPrefixLength));
+}
+
+static QString getQueryItemFromLink(const QUrl &url, const QString &queryItemName)
+{
+    QUrlQuery query(url);
+    if (!query.hasQueryItem(queryItemName)) {
+        return {};
+    }
+    return query.queryItemValue(queryItemName);
+}
+
+static QString getSinceQueryItemFromLink(const QUrl &url)
+{
+    return getQueryItemFromLink(url, QStringLiteral("since"));
+}
+
+void OcsActivityJob::processNextPage(const QNetworkReply *reply)
+{
+    if (reply->hasRawHeader("Link")) {
+        const auto nextLink = parseLinkFromHeader(reply->rawHeader("Link"));
+        if (!nextLink.isValid()) {
+            qCWarning(lcActivityJob) << "Link to next page submitted, but could not be parsed";
+            emit error();
+        }
+        const auto since = getSinceQueryItemFromLink(nextLink);
+        if (since.isEmpty()) {
+            qCWarning(lcActivityJob) << "Link to next page submitted, but the link has no since query item";
+            emit error();
+        }
+        startJsonApiJob(since);
+    }
```

Comment:
@FlexW I wonder if any emit (like `finished()`) is needed in case it was the last page and there is no Link header?

Reply:
No, thats not needed. This function only checks if there is a new page, and then starts a new round. The "finished" signal is connected in `startJsonApiJob`.